File: clients/scala/src/main/scala/ai_cup_22/model/Projectile.scala
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Weapon projectile
 *
 * @param id Unique id
 * @param weaponTypeIndex Index of the weapon this projectile was shot from (starts with 0)
 * @param shooterId Id of unit who made the shot
 * @param shooterPlayerId Id of player (team), whose unit made the shot
 * @param position Current position
 * @param velocity Projectile's velocity
 * @param lifeTime Left time of projectile's life
 */
case class Projectile(id: Int, weaponTypeIndex: Int, shooterId: Int, shooterPlayerId: Int, position: ai_cup_22.model.Vec2, velocity: ai_cup_22.model.Vec2, lifeTime: Double) {
    /**
     * Write Projectile to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeInt(stream, id)
        StreamUtil.writeInt(stream, weaponTypeIndex)
        StreamUtil.writeInt(stream, shooterId)
        StreamUtil.writeInt(stream, shooterPlayerId)
        position.writeTo(stream)
        velocity.writeTo(stream)
        StreamUtil.writeDouble(stream, lifeTime)
    }

    /**
     * Get string representation of Projectile
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Projectile { ")
        stringBuilder.append("id: ")
        stringBuilder.append(id)
        stringBuilder.append(", ")
        stringBuilder.append("weaponTypeIndex: ")
        stringBuilder.append(weaponTypeIndex)
        stringBuilder.append(", ")
        stringBuilder.append("shooterId: ")
        stringBuilder.append(shooterId)
        stringBuilder.append(", ")
        stringBuilder.append("shooterPlayerId: ")
        stringBuilder.append(shooterPlayerId)
        stringBuilder.append(", ")
        stringBuilder.append("position: ")
        stringBuilder.append(position)
        stringBuilder.append(", ")
        stringBuilder.append("velocity: ")
        stringBuilder.append(velocity)
        stringBuilder.append(", ")
        stringBuilder.append("lifeTime: ")
        stringBuilder.append(lifeTime)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Projectile {
    /**
     * Read Projectile from input stream
     */
    def readFrom(stream: java.io.InputStream): Projectile = Projectile(
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        StreamUtil.readInt(stream),
        ai_cup_22.model.Vec2.readFrom(stream),
        ai_cup_22.model.Vec2.readFrom(stream),
        StreamUtil.readDouble(stream)
    )
}

File: clients/swift/Sources/AiCup22/Model/Projectile.swift
/// Weapon projectile
public struct Projectile {
    /// Unique id
    let id: Int32

    /// Index of the weapon this projectile was shot from (starts with 0)
    let weaponTypeIndex: Int32

    /// Id of unit who made the shot
    let shooterId: Int32

    /// Id of player (team), whose unit made the shot
    let shooterPlayerId: Int32

    /// Current position
    let position: Vec2

    /// Projectile's velocity
    let velocity: Vec2

    /// Left time of projectile's life
    let lifeTime: Double

    /// Read Projectile from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Projectile {
        var id: Int32
        id = stream.readInt32()
        var weaponTypeIndex: Int32
        weaponTypeIndex = stream.readInt32()
        var shooterId: Int32
        shooterId = stream.readInt32()
        var shooterPlayerId: Int32
        shooterPlayerId = stream.readInt32()
        var position: Vec2
        position = Vec2.readFrom(stream)
        var velocity: Vec2
        velocity = Vec2.readFrom(stream)
        var lifeTime: Double
        lifeTime = stream.readDouble()
        return Projectile(id: id, weaponTypeIndex: weaponTypeIndex, shooterId: shooterId, shooterPlayerId: shooterPlayerId, position: position, velocity: velocity, lifeTime: lifeTime)
    }

    /// Write Projectile to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(id)
        stream.writeInt32(weaponTypeIndex)
        stream.writeInt32(shooterId)
        stream.writeInt32(shooterPlayerId)
        position.writeTo(stream)
        velocity.writeTo(stream)
        stream.writeDouble(lifeTime)
    }
}
